### DISCUSSION TITLE
Linking against Windows API libraries required by static libcurl

### DISCUSSION THREAD

**Discussion by mloskot** [valhalla/valhalla#3105] [General]
Spawning this thread off the Boost related issue here https://github.com/valhalla/valhalla/issues/2799#issuecomment-847637156 to avoid off-topic comments there.

@nilsnolde 
> If the CMake cache is there I always observe the same, not even half the reporting of a clean cmake run

Another example of a 'silent' cache'ed-out dependency is the CURL, for which CMake re-runs leave no trace:

```
CMake suite maintained and supported by Kitware (kitware.com/cmake).
-- The CXX compiler identification is MSVC 19.28.29914.0
-- The C compiler identification is MSVC 19.28.29914.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring in release mode
-- Found ZLIB: D:/a/1/s/vcpkg/installed/x64-windows/lib/zlib.lib (found version "1.2.11") 
-- Found ZLIB: D:/a/1/s/vcpkg/installed/x64-windows/lib/zlib.lib (found suitable version "1.2.11", minimum required is "1") 
-- Using pbf headers from D:/a/1/s/vcpkg/installed/x64-windows/include
-- Using pbf libs from D:/a/1/s/vcpkg/installed/x64-windows/lib/libprotobuf.lib
-- Using pbf release libs from 
-- Using pbf debug libs from D:/a/1/s/vcpkg/installed/x64-windows/lib/libprotobuf.lib
-- Using pbf-lite
-- Found SQLite3: D:/a/1/s/vcpkg/installed/x64-windows/lib/sqlite3.lib
-- Found Python: C:/hostedtoolcache/windows/Python/3.9.5/x64/python3.exe (found version "3.9.5") found components: Development Interpreter Development.Module Development.Embed 
-- pybind11 v2.6.2 
-- Installing python modules to C:\hostedtoolcache\windows\Python\3.9.5\x64\Lib\site-packages
-- Found GEOS...
-- GEOS_INCLUDE_DIR=D:/a/1/s/vcpkg/installed/x64-windows/include
-- GEOS_LIBRARY=D:/a/1/s/vcpkg/installed/x64-windows/lib/geos.lib;D:/a/1/s/vcpkg/installed/x64-windows/lib/geos_c.lib
-- Configuring done
-- Generating done
-- Build files have been written to: D:/a/build
```

I wonder, where do we link against the required Windows API libraries required by the static libcurl declared here:

https://github.com/valhalla/valhalla/blob/f005efa91e85e8f1df10e28c041ea8d35bed477c/CMakeLists.txt#L142-L144

I'd have expected we must link the tools like this:

```cmake
target_link_libraries(valhalla_run_route
  PRIVATE
    ${CURL_LIBRARY}
    $<$<CXX_COMPILER_ID:MSVC>:ws2_32>
    $<$<CXX_COMPILER_ID:MSVC>:wldap32>
    $<$<CXX_COMPILER_ID:MSVC>:advapi32>
    $<$<CXX_COMPILER_ID:MSVC>:crypt32>
    $<$<CXX_COMPILER_ID:MSVC>:normaliz>)
```

I'm very confident that on Windows the static libcurl requires those (all or some, depending on libcurl build config itself) , but I'm confused where does that happens in the Valhalla build, on Windows CI.

**Comment by kevinkreiser** [comment]
i'm not a cmake expert but i had assumed the call to `find_package` filled out all the relevant libraries to be linked for curl and shoved them into a variable that we used in the link line below. basically a combination between:

https://github.com/valhalla/valhalla/blob/f005efa91e85e8f1df10e28c041ea8d35bed477c/CMakeLists.txt#L140
and
https://github.com/valhalla/valhalla/blob/f005efa91e85e8f1df10e28c041ea8d35bed477c/CMakeLists.txt#L148

i have no clue though tbh. we could debug it with some liberal printing :smile:

**Reply by mloskot** [reply]
>  i had assumed the call to find_package filled out all the relevant libraries to be linked for curl and shoved them into a variable 

Looking at the current [FindCURL.cmake](https://github.com/Kitware/CMake/blob/8fea95319bcaac741faf93bd66b3a1838b56a418/Modules/FindCURL.cmake) I'm not finding anything like such magic to happen.

> we could debug it with some liberal printing 😄

I'll run some experiments and learn a bit more myself, stay tuned :-)
